pluginutil/pluginutil.go

```diff
@@ -0,0 +1,197 @@
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	gp "github.com/hashicorp/go-plugin"
+)
+
+type (
+	// InmemCreationFunc is a function that, when run, returns the thing you
+	// want created (almost certainly an interface that is also supported by a
+	// go-plugin plugin implementation)
+	InmemCreationFunc func() (interface{}, error)
+
+	// PluginClientCreationFunc is a function that, when run, returns a client
+	// corresponding to a spun out go-plugin plugin. The string argument is the
+	// filename. WithSecureConfig is supported as an option that will be round
+	// tripped to the given function if provided to this package so that it can
+	// be given to go-plugin.
+	PluginClientCreationFunc func(string, ...Option) (*gp.Client, error)
+)
+
+// PluginInfo contains plugin instantiation information for a single plugin,
+// parsed from the various maps and FSes that can be input to the BuildPluginMap
+// function.
+type PluginInfo struct {
+	ContainerFs              fs.FS
+	Filename                 string
+	InmemCreationFunc        InmemCreationFunc
+	PluginClientCreationFunc PluginClientCreationFunc
+}
+
+// BuildPluginMap takes in options that contain one or more sets of plugin maps
+// or filesystems and builds an overall mapping of a plugin name to its
+// information. The desired plugin can then be sent to CreatePlugin to actually
+// instantiate it. If a plugin is specified by name multiple times in option,
+// the last one wins.
+func BuildPluginMap(opt ...Option) (map[string]*PluginInfo, error) {
+	opts, err := GetOpts(opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing plugin options: %w", err)
+	}
+
+	if len(opts.withPluginSources) == 0 {
+		return nil, fmt.Errorf("no plugins available")
+	}
+
+	pluginMap := map[string]*PluginInfo{}
+	for _, sourceInfo := range opts.withPluginSources {
+		switch {
+		case sourceInfo.pluginFs != nil:
+			if opts.withPluginClientCreationFunc == nil {
+				return nil, fmt.Errorf("non-in-memory plugin found but no creation func provided")
+			}
+			dirs, err := fs.ReadDir(sourceInfo.pluginFs, ".")
+			if err != nil {
+				return nil, fmt.Errorf("error scanning plugins: %w", err)
+			}
+			// Store a match between the config type string and the expected plugin name
+			for _, entry := range dirs {
+				pluginType := strings.TrimSuffix(strings.TrimPrefix(entry.Name(), sourceInfo.pluginFsPrefix), ".gz")
+				if runtime.GOOS == "windows" {
+					pluginType = strings.TrimSuffix(pluginType, ".exe")
+				}
+				pluginMap[pluginType] = &PluginInfo{
+					ContainerFs:              sourceInfo.pluginFs,
+					Filename:                 entry.Name(),
+					PluginClientCreationFunc: opts.withPluginClientCreationFunc,
+				}
+			}
+		case sourceInfo.pluginMap != nil:
+			for k, creationFunc := range sourceInfo.pluginMap {
+				pluginMap[k] = &PluginInfo{InmemCreationFunc: creationFunc}
+			}
+		}
+	}
+
+	return pluginMap, nil
+}
+
+// CreatePlugin instantiates a given plugin either via an in-memory function or
```

Comment:
Would you mind documenting the need to call the cleanup() func returned... otherwise the caller has to read the code to figure out what's returned.

Reply:
Sure